### PR TITLE
checker: fix remove pending peer by replacing peer 

### PR DESCRIPTION
**Pull request by rleungx**
# What problem does this PR solve? 
Currently, once we remove a pending peer when a store is offline or down, there might be a gap before making a replica, which may be very dangerous. 

### What is changed and how it works?
This PR fixes this problem by replacing the pending peer instead of removing the peer directly.

### Check List 

Tests 

 - Unit test

Related changes

 - Need to cherry-pick to the release branch
 - Need to be included in the release notes

### CODE REVIEW

File: server/checker/replica_checker.go [hunk 1]
```diff
@@ -250,33 +250,34 @@ func (r *ReplicaChecker) fixPeer(region *core.RegionInfo, peer *metapb.Peer, sta
...
+	}
+
+	replacePending := fmt.Sprintf("replace-pending-%s-replica", status)
 	// Consider we have 3 peers (A, B, C), we set the store that contains C to
 	// offline/down while C is pending. If we generate an operator that adds a replica
 	// D then removes C, D will not be successfully added util C is normal again.
 	// So it's better to remove C directly.
 	if region.GetPendingPeer(peer.GetId()) != nil {
-		op, err := schedule.CreateRemovePeerOperator(removePending, r.cluster, schedule.OpReplica, region, peer.GetStoreId())
+		op, err := schedule.CreateMovePendingPeerOperator(replacePending, r.cluster, region, schedule.OpReplica, peer.GetStoreId(), newPeer.GetStoreId(), newPeer.GetId())
```
Comment (disksing) [inline server/checker/replica_checker.go]: Need to update ` So it's better to remove C directly` above?

File: server/checker/replica_checker_test.go [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+func TestReplicaChecker(t *testing.T) {
+	TestingT(t)
+}
+
+var _ = Suite(&testReplicaCheckerSuite{})
+
+type testReplicaCheckerSuite struct {
+	cluster *mockcluster.Cluster
+	rc      *ReplicaChecker
+	regions []*core.RegionInfo
```
Comment (disksing) [inline server/checker/replica_checker_test.go]: `*core.RegionInfo` should be enough. Only 1 region is used.

File: server/checker/replica_checker.go [hunk 3]
```diff
@@ -250,33 +250,20 @@ func (r *ReplicaChecker) fixPeer(region *core.RegionInfo, peer *metapb.Peer, sta
...
-	}
-
 	storeID, _ := r.SelectBestReplacementStore(region, peer, schedule.NewStorageThresholdFilter())
 	if storeID == 0 {
 		log.Debug("no best store to add replica", zap.Uint64("region-id", region.GetID()))
 		return nil
 	}
-	newPeer, err := r.cluster.AllocPeer(storeID)
-	if err != nil {
+	newPeer, e := r.cluster.AllocPeer(storeID)
```
Comment (shafreeck) [inline server/checker/replica_checker.go]: Why change it to `e`?

**Comment by rleungx** [comment]
I completely remove the pending logic when making replicas. When we have the pending peer, we still add another peer first to make the whole process safer. Is this reasonable? @disksing @nolouch

**Comment by disksing** [comment]
The operator could be blocked for a very long time if the pending peer does not recover.

**Comment by rleungx** [comment]
Since we already have learner, maybe it won't block anymore? @disksing

**Comment by disksing** [comment]
em, sounds make sense.

**Comment by disksing** [comment]
Should we remove the `enable learner` option?

**Comment by rleungx** [comment]
@disksing Yes. I think we can remove it.
/cc @nolouch What your opinion?

**Pull request merged**